### ISSUE TITLE
Allow GraphQL service components to query via GraphiQL in Definitions screen of API Docs

### ISSUE THREAD

**Issue by taras** [backstage/backstage#12392]
## Context

<img width="1321" alt="image" src="https://user-images.githubusercontent.com/74687/176909602-a4d29b4a-c7df-401c-8fba-228f9ccfa2c6.png">

I did some digging into this problem and I found a way to reproduce it

1. Go to https://demo.backstage.io/catalog/default/api/starwars-graphql/definition
2. Add the following query and hit play button
  ```graphql
  {
    allFilms {
      edges {
        node {
          id
      }
    }
  }
}
```
3. You'll get the following error.

```
{
  "errors": {
    "message": "Cannot use 'in' operator to search for 'subscribe' in null",
    "stack": "TypeError: Cannot use 'in' operator to search for 'subscribe' in null\n    at $r (https://demo.backstage.io/static/550.ae2c2479.chunk.js:80:297)\n    at https://demo.backstage.io/static/550.ae2c2479.chunk.js:49:169"
  }
}
```

This error is actually GraphiQL's attempt to parse response from the [default fetcher](https://github.com/backstage/backstage/blob/master/plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx#L55).

This is happening because `GraphQlDefinition` doesn't use the fetchers that are configured in [apis.tsx](https://github.com/backstage/demo/blob/master/packages/app/src/apis.ts#L49-L71). 

## Feature Suggestion

My understanding of the use case here is to allow users to query the API. Api Docs seems to be setup to use GraphiQL for documentation but not for querying. GraphiQL plugin can query using fetchers. 

## Possible Implementation

We could allow querying from GraphQLDefinitionWidget which would require providing a way to reuse or configure fetchers. Alternatively, we could change how we should documentation for GraphQL to no use GraphiQL.

**Comment by andrewthauer** [comment]
There may be some confusion here and I completely understand why. The definition in `apis.ts` for `graphQlBrowseApiRef` is meant for the `@backstage/plugin-graphiql` plugin. Although the `@backstage/plugin-api-docs` plugin uses the same component (https://github.com/graphql/graphiql), they are not connected in any way.

Introspection and querying would be nice in the `api-docs` version. However, discussed in #5590, doing so in a generic way that works across all catalog api's and org envs might be tricky to cover sufficiently. That said, having a programmatic interface similar or possibly shared across the `graphiql` and `api-docs` plugins might be the way to go here. There could be a shared package with the graphql endpoint config, etc. which requires programmatic wiring in the app or backend (possibly proxy in some cases?). There would need to be some gluing metadata to match a catalog api entity to an appropriate endpoint config though.

**Comment by taras** [comment]
I think this issue should be reopened because users keep bumping into it.

**Comment by andrewthauer** [comment]
Our team is going to be looking into the GraphQL docs situation internally starting next week. There might be something we can contribute back once we have a closer look at this.

@taras - Do you have any particular ideas/thoughts on this? In my mind there are 2 things here. There is documentation (usually best done with static docs generators) & there is the interactive usage of the schemas (typically through introspection). Graphiql v2 was released and is much more modular so was thinking of investigating what it offers over the current v1. Beyond that, it's an open road.

**Comment by taras** [comment]
@andrewthauer I've been keeping an eye on GraphQL documentation options. I haven't found anything. Did your look at the new GraphiQL docs yield anything?

**Comment by andrewthauer** [comment]
> @andrewthauer I've been keeping an eye on GraphQL documentation options. I haven't found anything. Did your look at the new GraphiQL docs yield anything?

@taras - Unfortunately, plans changed and we haven't looked into this further. For pure GraphQL documentation there is now https://magidoc.js.org. However, it's not a react component so it would likely require building static sites in CI, etc. and then having a link out to the static site. However, I haven't tried it out so perhaps there might be a nicer way. I kind of thing there are 2 personas here. One where someone wants to interact with the GraphQL schema using introspection and the other is a more traditional API documentation approach. Perhaps there is a new for both given the GraphQL SDL itself isn't really tailored to do both really well (or at all) like OpenAPI can.

**Comment by jtreher** [comment]
Did this ever make it into the RFC? I do agree that it's currently confusion that it uses Graphiql to give the illusion that your query is going to work. I'd almost rather see it just render the schema in a decent format if getting it to run actual queries is too big of a hill to climb.

**Comment by andrewthauer** [comment]
> Did this ever make it into the RFC? I do agree that it's currently confusion that it uses Graphiql to give the illusion that your query is going to work. I'd almost rather see it just render the schema in a decent format if getting it to run actual queries is too big of a hill to climb.

I don't think there is an official RFC, but there have been a few related GH issues on the general subject and improvements of the GraphQL API Docs experience. I think we should probably split this into a couple of issues? For example:

a) Remove the confusing GraphiQL query building and execution and just have a pure "docs" view.

I started playing around with `GraphiQL` v2 and it appears you can compose just the docs part without the rest of the editor. However, it's a bit sparse and looks funny initially. Maybe some styling would help. Another option is to use something like https://magidoc.js.org/ which is purpose built for this, but since it's a static site generator it doesn't integrate with react in any way and would either need build & hosted externally and linked to.

b) Allow API Docs of type `graphql` to be interactive similar to OpenAPI

Since the GraphQL SDL spec doesn't support these annotations out of the box we'd need to define a way to do this. Possibly options are through custom annotations on the SDL, having an annotation configuration on the related entity to define the connection, through code, through app-config.yaml or a combination. Technically not hard, but a proposal would be required to handle appropriate uses cases and requirements (connection types, auth, etc.)

**Comment by jtreher** [comment]
@andrewthauer Thanks for your thoughtful feedback. I like A for now. I don't know about you, but our GraphQL Api's expose a playground off each API and there isn't a place to drop that link on the definition page. So, a definition page that 1) links to the playground for the API (if available) and lists out the spec for browsing would clear up any confusion.

**Comment by andrewthauer** [comment]
So we already have the `Raw` view of the SDL. Would be really nice if it was syntax highlighted though. Might be something to also investigate. Few potentially immediate options I see are:

- Replace the `graphiql` page with a component that can read some annotation that links to a dedicated docs site (e.g. `graphql.org/docs-site-url: https://some.place.hosting/graphql/api/`
- Look at upgrading to `graphiql` v2 and only showing the `Docs` part.
- A combination of the 2 or possibly providing a pick and choose model between the existing view and a new view.

Technically removing the `graphiql` view and adding support for a custom `annotation` shouldn't be too difficult. The challenge being this would be a breaking change from a feature/user perspective potentially. Granted it might not be that big of a deal. Technically you can override the entire component now, so another approach could be to provide a new component that links to docs as the default or optionally configurable to replace the current view.

I'd love to contribute this, but given my focus, I definitely won't have the time in the short to medium term.

**Comment by SujithChowdari** [comment]
Any alternative solutions to this? I would like to see our developers running queries from the GraphiQL interface rather than using that interface for just showing the schema

**Comment by andrewthauer** [comment]
Since the GraphQL SDL itself does not contain provisions for auth, endpoints etc like OpenAPI this would require something custom which entities themselves may need to provide this information or be coded outside the API spec. Might be worth looking at the new [graphql-voyager](https://github.com/backstage/backstage/tree/master/plugins/graphql-voyager) plugin. I haven't tried it myself but am curious if it could be adapted to the API docs plugin in a more general way. I think the problem again being that it requires introspection and this could require auth which could need to be handled differently for each API.

**Comment by taras** [comment]
@andrewthauer the voyager might be a good option. I was setting it up earlier and there is one wrinkle. It requires `voyager.worker.js` to be added to the `public` directory. It's in node_modules. I don't know if there is an API for adding files from node_modules to public directory from a plugin.

**Comment by shrutiyer** [comment]
@taras @andrewthauer seeing that this issue has gone stale and closed, is there a different plan for GraphQL APIs to be executed from within Backstage?

**Comment by taras** [comment]
@shrutiyer do you mean to experiment with a GraphQL API in API Docs?

**Comment by shrutiyer** [comment]
Correct. Similar to the "Try it out" feature within swagger/openapi specs. Is there any feature planned to do the same for GraphQL APIs?

**Comment by taras** [comment]
Part of the challenge is that it's unclear how this should be solved. Each backend usually needs credentials and we don't have a way to provide these. We discussed remove the "try it" and replacing it with just documentation but it wasn't a priority for anyone.

**Comment by shalinichennoju** [comment]
@taras @andrewthauer Seems we cannot query the graphql apis in backstage but only can view the schema with the existing code??                       if I integrate @backstage/plugin-graphiql, will this allow me to query the graphql endpoints from the backstage UI?

**Comment by andrewthauer** [comment]
@shalinichennoju - The `graphiql` plugin will let you query through introspection when configured for each endpoint. However, this is a separate UI and plugin from the `api-docs` plugin and not based on the catalog or API entities.

**Comment by younky-yang** [comment]
started to work on some enhancements for this one. First for plugin-graphiql, I added a function to GraphQLEndpoints.ts to allow it query all API with type=graphql, then create the endpoints automatically. This depends on a annotation for the API entity. (backstage.io/api-graphql-url). then in GraphiQLPage.ts, I call this function after the GraphQLBrowserAPI.getEndpoints(). 

With this approach, when I enabled graphiql plugin, I can play with the APIs that defined with API entities. 

another change I made is to replace the tab with Select and also set another annotation backstage.io/api-graphql-url-auth-url for remind user where to get the token for querying this graphql api. 

The last change I made is to upgrade the graphiql plugin to 3.0.5 which is the latest, I am still working on some details of this change, and would like to raise a PR for this enhancements.
